File: Inteligencia-Artificial_Programa_1/Puzzle/src/Ilayout.java
import java.util.List;

public interface Ilayout {
    /**
     @return the children of the receiver.
     */
    List<Ilayout> children();
    /**
     @return true if the receiver equals the argument l; return false otherwise.
     */
    boolean isGoal(Ilayout l);
    /**
     @return the cost from the receiver to a successor
     */
    double getK();
}

File: Inteligencia_Artificial_Programa_2C++/Ilayout.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>
#include <functional>
#include <memory>
// Define the Ilayout interface (equivalent to Java interface)
class Ilayout {
public:
    virtual double computeHeuristic(std::shared_ptr<Ilayout> goalLayout) = 0;
    virtual bool isGoal(std::shared_ptr<Ilayout> l) = 0;
    virtual std::vector<std::shared_ptr<Ilayout>> children() = 0;
    virtual double getK() const = 0;  // Movement cost
    virtual bool operator==(const Ilayout& other) const = 0;  // For equality check
    virtual std::size_t hashCode() const = 0;  // For hashing
    virtual ~Ilayout() {}
};